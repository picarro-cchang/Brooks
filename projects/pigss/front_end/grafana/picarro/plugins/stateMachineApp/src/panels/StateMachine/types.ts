import { any } from "prop-types";

export interface ButtonInfo {
  caption: string;
  className: string;
  response: string;
}

export interface CommandPanelOptions {
  uistatus: {
    [key: string]: string;
  };
  ws_sender: (o: object) => void;
  plan: Plan;
}

export interface ModalInfo {
  show: boolean;
  html: string;
  num_buttons: number;
  buttons: { [key: string]: ButtonInfo };
}

export interface Plan {
  max_steps: number;
  panel_to_show: number;
  current_step: number;
  focus: PlanFocus;
  last_step: number;
  steps: { [key: string]: PlanStep };
  num_plan_files: number;
  plan_files: { [key: string]: string };
  plan_filename: string;
  bank_names: {
    [key: number]: {
      name: string;
      channels: { [key: number]: string };
    };
  };
}

export interface PlanFocus {
  row: number;
  column: number;
}

export interface PlanLoadPanelOptions {
  plan: Plan;
  ws_sender: (o: object) => void;
  isChanged: boolean;
  updateFileName: (x: boolean) => void;
}

export interface PlanPanelOptions {
  uistatus: { [key: string]: string };
  plan: Plan;
  setFocus: (row: number, column: number) => void;
  ws_sender: (o: object) => void;
  isChanged: boolean;
  updateFileName: (x: boolean) => void;
  // addChanneltoPlan: (bank: number, channel: number) => void;
  bankAddition: {[key: string]: number}
}

export enum PlanPanelTypes {
  NONE = 0,
  PLAN = 1,
  LOAD = 2,
  SAVE = 3,
  EDIT = 4
}

export interface PlanSavePanelOptions {
  plan: Plan;
  ws_sender: (o: object) => void;
  isChanged: boolean;
  updateFileName: (x: boolean) => void;
}

export interface BankConfig {
  chan_mask: number;
  clean: number;
}

export interface PlanStep {
  banks: { [key: number]: BankConfig };
  reference: number;
  duration: number;
}

export interface EditPanelOptions {
  uistatus: {
    [key: string]: any;
  };
  ws_sender: (o: object) => void;
  plan: Plan;
}
