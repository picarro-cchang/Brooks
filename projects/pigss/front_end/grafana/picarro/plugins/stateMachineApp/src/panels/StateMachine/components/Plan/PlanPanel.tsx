import React, { Component, ReactText } from "react";
import ReactList from "react-list";
import { PlanPanelOptions, PlanStep, Plan } from "../types";
import { PlanService } from "../../api/PlanService";
import fs from "fs";

export interface State {
  refVisible: boolean;
  plan: Plan;
  isLoaded: boolean;
  bankAdditionClicked: {};
  fileName: string;
  fileNames: string[];
}

interface planStorage {
  planState: Plan;
}


//TODO: Add Validation
class PlanPanel extends Component<PlanPanelOptions, State> {
  constructor(props) {
    super(props);
    this.state = {
      bankAdditionClicked: {},
      isLoaded: false,
      refVisible: true,
      plan: this.props.plan,
      fileName: this.props.fileName,
      fileNames: []
    };
    this.addToPlan = this.addToPlan.bind(this);
    this.getAvailableBanks = this.getAvailableBanks.bind(this);
    this.updateFocus = this.updateFocus.bind(this);
    this.updateDuration = this.updateDuration.bind(this);
    this.updateCurrentStep = this.updateCurrentStep.bind(this);
    this.insertRow = this.insertRow.bind(this);
    this.clearPlan = this.clearPlan.bind(this);
  }

  validation(plan: Plan, checkAvailable=true) {
    if (plan.last_step <= 0) {
      this.props.setModalInfo(
        true, 
        `<h4 style='color:black'>Plan is Empty</h4>`,
        0,
        {},
        "misc")
      return false
    }
    if (plan.current_step > plan.last_step || plan.current_step < 1) {
      this.props.setModalInfo(
        true, 
        `<h4 style='color:black'>Plan Step must be between 1 and ${plan.last_step}</h4>`,
        0,
        {},
        "misc")
      return false
    }
    const minDuration = 20;
    const all_banks = this.getAvailableBanks();
    console.log(all_banks)
    for (let i = 0; i < plan.last_step; i++) {
      let row = i + 1;
      let s = plan.steps[row]
      if (!("duration" in s) || !("reference" in s) || !("banks" in s)) {
        this.props.setModalInfo(
          true, 
          `<h4 style='color:black'>Malformed Data.</h4>`,
          0,
          {},
          "misc")
        return false
      }
      if (s.duration <= 0 || s.duration < minDuration) {
        this.props.setModalInfo(
          true, 
          `<h4 style='color:black'>Invalid Duration at step ${row}. Duration must be greater then ${minDuration}</h4>`,
          0,
          {},
          "misc")
        return false
      }
      for(let bank in s.banks) {
        console.log(bank, "is included? ", all_banks.includes(bank))
        if (!(all_banks.includes(Number(bank)))) {
          this.props.setModalInfo(
            true, 
            `<h4 style='color:black'>Invalid Bank at Step ${row}</h4>`,
            0,
            {},
            "misc")
          return false
        }
        let bank_config = s.banks[bank];
        if (!("chan_mask" in bank_config) || !("clean" in bank_config)) {
          this.props.setModalInfo(
            true, 
            `<h4 style='color:black'>Invalid Data for Bank ${bank} in step ${row}</h4>`,
            0,
            {},
            "misc")
          return false
        }
        if (bank_config.chan_mask < 0 || bank_config.chan_mask > 256) {
          this.props.setModalInfo(
            true, 
            `<h4 style='color:black'>Invalid Channel Selection for Bank ${bank} in Step ${row}</h4>`,
            0,
            {},
            "misc")
          return false
        }
        if (checkAvailable) {
            const mask = bank_config.chan_mask;
            if (mask != 0) {
              const channel = (mask & -mask).toString(2).length;
              const channelBtn = this.props.uistatus.channel[bank][channel]
              if (channelBtn == "DISABLED") {
                console.log(channelBtn, " Does not equal AVAILABLE")
                this.props.setModalInfo(
                true, 
                `<h4 style='color:black'>Unavailable Port at Step ${row}</h4>`,
                0,
                {},
                "misc")
                return false
              }
              
            }
        }
      }
    }
    console.log('all done')
    return true
  }


  shouldComponentUpdate(nextProps) {
    if (this.props.bankAddition !== nextProps.bankAddition) {
      this.setState({
        bankAdditionClicked: nextProps.bankAddition
      });
      if (nextProps.bankAddition.bank != undefined) {
        const bank = nextProps.bankAddition.bank;
        const channel = nextProps.bankAddition.channel;
        const row = this.state.plan.focus.row;
        this.changeChannelstoProperFormat(bank, channel, row);
      }
    }
    return true;
  }

  saveFile() {
    this.props.updatePanel(2);
    this.props.updateSavedFileState(this.state.plan);
  }

  isSaveDisabled() {
    if(this.state.plan.focus.row > this.state.plan.last_step) {
      return true
    } else if (!this.props.isEdited) {
      return true
    } else if (this.state.plan.plan_filename == "") {
      return true
    }
    return false
  }

  overwriteFile() {
    const valid = this.validation(this.state.plan, true)
    if (valid) {
      this.props.setModalInfo(true, `<div>Save file as ${this.state.fileName}? This will overwrite the file.</div>`, 2, {
      1: {
        caption: "Save",
        className: "btn btn-success btn-large",
        response: {plan: this.state.plan}
      },
      2: {
        caption: "Cancel",
        className: "btn btn-danger btn-large",
        response: null
      }
      }, 'saveOverwrite')
    }
  }

  clearPlan() {
    this.props.updateFileName(true);
    let plan = {...this.state.plan};
    plan.steps = {};
    plan.current_step = 1;
    plan.focus = {
      row: 1, 
      column: 1
    }
    plan.last_step = 0;
    this.setState({plan})
  }

  deleteRow() {
    let row = this.state.plan.focus.row;
    let col = this.state.plan.focus.column;
    let num_steps = this.state.plan.last_step;
    const plan = { ...this.state.plan };
    if (num_steps > 0 && row <= num_steps) {
      for(let r = row; r< this.state.plan.last_step; r += 1) {
        let s = this.state.plan.steps[r + 1];
        plan.steps[r] = s;
      }
      plan.last_step = num_steps - 1;
      if (row == 1) {
        col = 1
      } else {
        row = row - 1
      }
      plan.focus = {
        row,
        column: col
      }
      this.setState({ plan });
    }
    this.props.updateFileName(true)
  }

  getAvailableBanks() {
    const availableBanks = [];
    if (("bank" in this.props.uistatus) as any) {
      for (let i = 1; i <= 4; i++) {
        if ((this.props.uistatus as any).bank.hasOwnProperty(i)) {
          availableBanks.push(i);
        }
      }
    }
    return availableBanks;
  }

  addToPlan(bank_config, reference: number) {
    this.props.updateFileName(true);
    let row = this.state.plan.focus.row;
    const col = this.state.plan.focus.column;
    let duration;
    if (row >= this.state.plan.max_steps && col == 2) {
      // error
      console.log("Max Number of Steps Reached!")
    }
    if (col == 2) {
      row += 1;
    }
    if (row <= this.state.plan.last_step) {
      duration = this.state.plan.steps[row].duration;
    } else {
      duration = 0;
    }

    const bank_step = {
      banks: bank_config,
      reference,
      duration
    };

    const plan = { ...this.state.plan };
    plan.steps[row] = bank_step;

    if (this.state.plan.last_step < row) {
      plan.last_step = row;
    }
    plan.focus = {
      row,
      column: 2
    };
    this.setState({ plan });
  }

  updateFocus(row: number, column: number) {
    const plan = { ...this.state.plan };
    if (row <= this.state.plan.last_step || row == this.state.plan.last_step) {
      plan.focus = {
        row,
        column
      };
      this.setState({ plan });
    }
  }

  updateDuration(row: number, duration: number) {
    this.props.updateFileName(true);
    const plan = { ...this.state.plan };
    plan.steps[row].duration = duration;
    this.setState({ plan });
  }

  updateCurrentStep(row: number) {
    this.props.updateFileName(true);
    const plan = { ...this.state.plan };
    plan.current_step = row;
    this.setState({ plan });
  }

  insertRow() {
    const all_banks = this.getAvailableBanks();
    const row = this.state.plan.focus.row;
    const col = this.state.plan.focus.column;
    const num_steps = this.state.plan.last_step;
    const plan = { ...this.state.plan };
    if (num_steps < this.state.plan.max_steps) {
      for (let i = this.state.plan.last_step; i > row - 1; i -= 1) {
        const s = this.state.plan.steps[i];
        // insert the plan 1 row down
        plan.steps[i + 1] = s;
      }
      const bank_config = {};
      for (const b in all_banks) {
        bank_config[all_banks[b]] = { clean: 0, chan_mask: 0 };
      }
      plan.steps[row] = {
        banks: bank_config,
        reference: 0,
        duration: 0
      };
      plan.last_step = num_steps + 1;
    }
    plan.focus = {
      row,
      column: col
    };
    this.setState({ plan });
    this.props.updateFileName(true)
  }

  changeChannelstoProperFormat(bank, channel, row) {
    const all_banks = this.getAvailableBanks();
    const bank_config = {};
    for (const b in all_banks) {
      bank_config[all_banks[b]] = { clean: 0, chan_mask: 0 };
    }
    if (bank == 0) {
      this.addToPlan(bank_config, 1);
    } else {
      if (channel == 0) {
        bank_config[bank].clean = 1;
        this.addToPlan(bank_config, 0);
      } else {
        bank_config[bank].chan_mask = 1 << (channel - 1);
        this.addToPlan(bank_config, 0);
      }
    }
  }

  exitPlanPanel() {
    this.props.setModalInfo(true, `<h4 style='color: black'/>Are you sure you want to leave? Any unsaved changes will be lost.</h4>`, 2,
    {
      1: {
        caption: "Continue",
        className: "btn btn-success btn-large",
        response: 'exit'
      },
      2: {
        caption: "Cancel",
        className: "btn btn-danger btn-large",
        response: null
      }
    }, 'exitPlan')
    
  }

  focusComponent: any = null;
  focusTimer: any = null;

  manageFocus = (component: any) => {
    // Manages focus by storing the component which is to receive focus
    //  and calling its focus() method after a short timeout. Multiple
    //  calls to manageFocus during the expiry of the timer will cause
    //  only the last component to receive the focus. This prevents
    //  oscillations in which a cycle of components receive focus in
    //  quick succession.
    if (this.focusTimer !== null) {
      clearTimeout(this.focusTimer);
    }
    this.focusComponent = component;
    this.focusTimer = setTimeout(() => {
      this.focusComponent.focus();
      this.focusTimer = null;
    }, 200);
  };

  makePlanRow = (row: number) => {
    let portString = "";
    let durationString = "";
    if (this.state.plan.last_step >= row) {
      const planRow = this.state.plan.steps[row] as PlanStep;
      if (planRow.reference != 0) {
        portString = "Reference";
      } else {
        for (const bank in planRow.banks) {
          if (planRow.banks.hasOwnProperty(bank)) {
            const bank_name = this.state.plan.bank_names[bank].name;
            const bank_config = planRow.banks[bank];
            if (bank_config.clean != 0) {
              portString = `Clean ${bank_name}`;
              break;
            } else if (bank_config.chan_mask != 0) {

              const mask = bank_config.chan_mask;
              // Find index of first set bit using bit-twiddling hack
              const channel = (mask & -mask).toString(2).length;
              const ch_name = this.state.plan.bank_names[bank].channels[
                channel
              ];
              let bankNum = Number(bank)
              let portNumber = (bankNum - 1)*8 + channel
              portString =  portNumber+ ": " + ch_name;
              break;
            }
          }
        }
      }
      if (planRow.duration !== 0) {
        durationString = `${planRow.duration}`;
      }
    }

    return (
      <div className="gf-form-inline" key={row}>
        {row < 10 ? (
          <label
            className="col-sm-1 panel-plan-text"
            style={{ fontSize: "17px", marginTop: "5px" }}
          >
            {row + ". "}
          </label>
        ) : (
          <label
            className="col-sm-1 panel-plan-text"
            style={{
              marginLeft: "-7px",
              marginRight: "7px",
              fontSize: "17px",
              marginTop: "5px"
            }}
          >
            {row + ". "}
          </label>
        )}
        <div className="col-sm-6 col-bank-name">
          <input
            ref={input =>
              input &&
              this.state.plan.focus.row === row &&
              this.state.plan.focus.column === 1 &&
              this.manageFocus(input)
            }
            type="text"
            className="form-control plan-input panel-plan-text"
            id={"plan-port-" + row}
            onFocus={e => {
              // this.changeChannelstoProperFormat(bank, channel, row);
              this.updateFocus(row, 1);
            }}
            onChange={e => {
              this.props.updateFileName(true);
            }}
            style={{ maxWidth: "90%", float: "left", marginLeft: "2px" }}
            value={portString}
            placeholder="Select port"
          />
        </div>
        <div className="col-sm-3" style={{ paddingLeft: "0px" }}>
          <input
            ref={input =>
              input &&
              this.state.plan.focus.row === row &&
              this.state.plan.focus.column === 2 &&
              this.manageFocus(input)
            }
            onChange={e => {
              this.props.updateFileName(true);
              this.updateDuration(row, Number(e.target.value));
            }}
            onFocus={e => {
              this.updateFocus(row, 2);
            }}
            maxLength={8}
            minLength={1}
            type="text"
            className="form-control input-small plan-input panel-plan-text"
            id={"plan-duration-" + row}
            style={{ maxWidth: "100%" }}
            value={durationString}
            placeholder="Duration"
          />
        </div>
        <label
          className="panel-plan-text"
          style={{ marginLeft: "-15px", paddingRight: "5px" }}
        >
          s
        </label>
        <label className="col-sm-1 radio-btn">
          <input
            type="radio"
            id={"plan-row-" + row}
            checked={row == this.state.plan.current_step}
            onChange={e => this.updateCurrentStep(row)}
            style={{ maxWidth: "100%" }}
          />
          <span className="checkmark"></span>
        </label>
      </div>
    );
  };

  renderItem = (index: number, key: ReactText) => (
    <div key={key}>{this.makePlanRow(index + 1)}</div>
  );

  render() {
    console.log("PLAN STATE PLAN PANEL ", this.state.plan)
    const file_name = this.state.plan.plan_filename; ///TROWS ERROR cannot read property plan_filename of null, which means this.state.plan is null 
    return (
      <div>
        <div className="panel-plan">
          <h2 className="panel-plan-text">Schedule</h2>
          <span
            className="cancel panel-plan-text"
            id="cancel-x"
            onClick={e => {
              if (this.props.isEdited) {
                this.exitPlanPanel();
              } else {
                this.props.layoutSwitch();
              }
            }}
          ></span>
          <h6 className="panel-plan-text">
            Please click on available channels to set up a schedule, then click
            on the radio button to select starting position.
          </h6>
          {this.state.plan.plan_filename && !this.props.isEdited ? (
            <div>
              <h6 className="panel-plan-text">
                Currently viewing File:{" "}
                <span style={{ color: "white" }}>{file_name}</span>
              </h6>
            </div>
          ) : (
            <h6 style={{ color: "white" }}>
              Currently not viewing a saved file
            </h6>
          )}
          <div className="panel-plan-inner">
            <form>
              <ReactList
                itemRenderer={this.renderItem}
                length={this.state.plan.max_steps}
                type={"uniform"}
              />
            </form>
          </div>
          <div className="container">
            <div className="row text-center btn-row-1">
              <div className="col-sm-3">
                <button
                  type="button"
                  id="insert-btn"
                  disabled={
                    this.state.plan.focus.row > this.state.plan.last_step
                  }
                  onClick={e => {
                    this.insertRow();
                    this.props.updateFileName(true);
                  }}
                  className={"btn btn-block btn-group"}
                >
                  Insert Row
                </button>
              </div>
              <div className="col-sm-3">
                <button
                  type="button"
                  id="save-btn"
                  disabled={
                    this.state.plan.focus.row > this.state.plan.last_step
                  }
                  onClick={e => {
                    this.saveFile();
                  }}
                  className={"btn btn-block btn-light btn-group"}
                >
                  Save As
                </button>
              </div>

              <div className="col-sm-3">
                <button
                  type="button"
                  id="load-btn"
                  onClick={e => this.props.updatePanel(1)}
                  className={"btn btn-block btn-light btn-group"}
                >
                  Load
                </button>
              </div>
            </div>
            <div className="row btn-row-2">
              <div className="col-sm-3">
                <button
                  type="button"
                  id="delete-btn"
                  disabled={
                    this.state.plan.focus.row > this.state.plan.last_step
                  }
                  onClick={e => {
                    this.props.updateFileName(true);
                    this.deleteRow();
                  }}
                  className={"btn btn-block btn-cancel btn-group"}
                >
                  Delete Row
                </button>
              </div>
              <div className="col-sm-3">
                <button
                  type="button"
                  id="clear-btn"
                  disabled={
                    this.state.plan.focus.row > this.state.plan.last_step
                  }
                  onClick={e => {
                    this.props.updateFileName(true);
                    this.clearPlan();
                  }}
                  className={"btn btn-block btn-cancel btn-group"}
                >
                  Clear
                </button>
              </div>

              <div className="col-sm-3">
                <button
                  type="button"
                  id="ok-btn"
                  disabled={this.isSaveDisabled()}
                  onClick={e => {
                    this.overwriteFile();
                    
                  }}
                  className={"btn btn-block btn-green btn-group"}
                >
                  Save
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

export default PlanPanel;
