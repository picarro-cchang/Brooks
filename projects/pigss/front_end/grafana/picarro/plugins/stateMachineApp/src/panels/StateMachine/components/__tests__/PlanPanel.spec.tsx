import React from 'react';
import {shallow, mount} from 'enzyme';
import 'jest-styled-components'
import WS from 'jest-websocket-mock';
import Modal from 'react-responsive-modal';
import PlanPanel from '../PlanPanel';
import {PlanPanelOptions, Plan} from '../../types';

const mockSetFocus = jest.fn();
const mockWSSender = jest.fn();
const mockUpdateFilename = jest.fn();

const defaultProps: PlanPanelOptions = {
    uistatus: {},
    plan: {
        max_steps: 32,
        panel_to_show: 3,
        current_step: 1,
        focus: {
            row: 1,
            column: 2
        },
        last_step: 1,
        steps: {
            "1": {
                "banks": {
                    "1": {
                    "clean": 0,
                    "chan_mask": 1
                    },
                    "2": {
                    "clean": 0,
                    "chan_mask": 0
                    }
                },
                "reference": 0,
                "duration": 30
            }
        },
        num_plan_files: 1,
        plan_filename: "test.pln",
        plan_files: {
            "1": "test.pln"
        },
        bank_names: {
            "1": {
                "name": "B1",
                "channels": {
                    "1": "Channel 1",
                    "2": "Channel 2",
                    "3": "Ch. 3",
                    "4": "Ch. 4",
                    "5": "Ch. 5",
                    "6": "Ch. 6",
                    "7": "Ch. 7",
                    "8": "Ch. 8"
                }
            },
            "2": {
                "name": "B1",
                "channels": {
                    "1": "Channel 1",
                    "2": "Channel 2",
                    "3": "Ch. 3",
                    "4": "Ch. 4",
                    "5": "Ch. 5",
                    "6": "Ch. 6",
                    "7": "Ch. 7",
                    "8": "Ch. 8"
                }
            },
            "3": {
                "name": "Bank 2",
                "channels": {
                    "1": "Ch. 1",
                    "2": "Ch. 2",
                    "3": "Ch. 3",
                    "4": "Ch. 4",
                    "5": "Ch. 5",
                    "6": "Ch. 6",
                    "7": "Ch. 7",
                    "8": "Ch. 8"
                }
            },
            "4": {
                "name": "Bank 3",
                "channels": {
                    "1": "Ch. 1",
                    "2": "Ch. 2",
                    "3": "Ch. 3",
                    "4": "Ch. 4",
                    "5": "Ch. 5",
                    "6": "Ch. 6",
                    "7": "Ch. 7",
                    "8": "Ch. 8"
                }
            },
        }
    },
    setFocus: mockSetFocus,
    ws_sender: mockWSSender,
    isChanged: false,
    updateFileName: mockUpdateFilename
};

describe('<PlanPanel />', () => {
    const wrapper = shallow(<PlanPanel {...defaultProps} />);
    const instance = wrapper.instance() as PlanPanel;

    it('Snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('makePlanRow', ()=> {
        const value = instance.makePlanRow(5);
        expect(value).toMatchSnapshot();
    });

    it('manageFocus', ()=> {
    });

    it('renderItem', ()=> {
        instance.makePlanRow = jest.fn();
        const listItem = instance.renderItem(1,1);
        expect(instance.makePlanRow).toHaveBeenCalled();
        expect(listItem).toMatchSnapshot();
    });

    it('ReactList renders', ()=> {
        const planList = mount(<PlanPanel {...defaultProps} />);
        expect(planList).toMatchSnapshot();
        const channelInput = planList.find('input').at(0)
        channelInput.simulate('keyDown');
        // expect(channelInput.props().onFocus).toHaveBeenCalled();
        // expect(mockWSSender).toHaveBeenCalled();
    });

    it('Insert Button', () => {

    });

    it('Save Button', () => {

    });

    it('Load Button', () => {

    });

    it('Delete Button', () => {

    });

    it('Clear Button', () => {

    });

    it('Ok Button', () => {

    });
});