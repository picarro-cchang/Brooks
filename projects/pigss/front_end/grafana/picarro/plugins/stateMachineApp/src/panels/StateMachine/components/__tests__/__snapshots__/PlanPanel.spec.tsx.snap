// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PlanPanel /> ReactList renders 1`] = `
<PlanPanel
  isChanged={false}
  plan={
    Object {
      "bank_names": Object {
        "1": Object {
          "channels": Object {
            "1": "Channel 1",
            "2": "Channel 2",
            "3": "Ch. 3",
            "4": "Ch. 4",
            "5": "Ch. 5",
            "6": "Ch. 6",
            "7": "Ch. 7",
            "8": "Ch. 8",
          },
          "name": "B1",
        },
        "2": Object {
          "channels": Object {
            "1": "Channel 1",
            "2": "Channel 2",
            "3": "Ch. 3",
            "4": "Ch. 4",
            "5": "Ch. 5",
            "6": "Ch. 6",
            "7": "Ch. 7",
            "8": "Ch. 8",
          },
          "name": "B1",
        },
        "3": Object {
          "channels": Object {
            "1": "Ch. 1",
            "2": "Ch. 2",
            "3": "Ch. 3",
            "4": "Ch. 4",
            "5": "Ch. 5",
            "6": "Ch. 6",
            "7": "Ch. 7",
            "8": "Ch. 8",
          },
          "name": "Bank 2",
        },
        "4": Object {
          "channels": Object {
            "1": "Ch. 1",
            "2": "Ch. 2",
            "3": "Ch. 3",
            "4": "Ch. 4",
            "5": "Ch. 5",
            "6": "Ch. 6",
            "7": "Ch. 7",
            "8": "Ch. 8",
          },
          "name": "Bank 3",
        },
      },
      "current_step": 1,
      "focus": Object {
        "column": 2,
        "row": 1,
      },
      "last_step": 1,
      "max_steps": 32,
      "num_plan_files": 1,
      "panel_to_show": 3,
      "plan_filename": "test.pln",
      "plan_files": Object {
        "1": "test.pln",
      },
      "steps": Object {
        "1": Object {
          "banks": Object {
            "1": Object {
              "chan_mask": 1,
              "clean": 0,
            },
            "2": Object {
              "chan_mask": 0,
              "clean": 0,
            },
          },
          "duration": 30,
          "reference": 0,
        },
      },
    }
  }
  setFocus={[MockFunction]}
  uistatus={Object {}}
  updateFileName={[MockFunction]}
  ws_sender={[MockFunction]}
>
  <div>
    <div
      className="panel-plan"
    >
      <h2
        className="panel-plan-text"
      >
        Schedule
      </h2>
      <span
        className="cancel panel-plan-text"
        onClick={[Function]}
      />
      <h6
        className="panel-plan-text"
      >
        Please click on available channels to set up a schedule, then click on the radio button to select starting position.
      </h6>
      <div>
        <h6
          className="panel-plan-text"
        >
          Currently viewing File:
           
          <span
            style={
              Object {
                "color": "white",
              }
            }
          >
            test.pln
          </span>
        </h6>
      </div>
      <div
        className="panel-plan-inner"
      >
        <form>
          <ReactList
            axis="y"
            itemRenderer={[Function]}
            itemsRenderer={[Function]}
            length={32}
            minSize={1}
            pageSize={10}
            scrollParentGetter={[Function]}
            scrollParentViewportSizeGetter={[Function]}
            threshold={100}
            type="uniform"
            useStaticSize={false}
            useTranslate3d={false}
          >
            <div
              style={
                Object {
                  "position": "relative",
                }
              }
            >
              <div
                style={
                  Object {
                    "WebkitTransform": "translate(0px, 0px)",
                    "msTransform": "translate(0px, 0px)",
                    "transform": "translate(0px, 0px)",
                  }
                }
              >
                <div>
                  <div
                    key="0"
                  >
                    <div
                      className="gf-form-inline"
                      key="1"
                    >
                      <label
                        className="col-sm-1 panel-plan-text"
                        style={
                          Object {
                            "fontSize": "17px",
                            "marginTop": "5px",
                          }
                        }
                      >
                        1. 
                      </label>
                      <div
                        className="col-sm-6 col-bank-name"
                      >
                        <input
                          className="form-control plan-input panel-plan-text"
                          id="plan-port-1"
                          onChange={[Function]}
                          onFocus={[Function]}
                          placeholder="Select port"
                          style={
                            Object {
                              "float": "left",
                              "marginLeft": "2px",
                              "maxWidth": "90%",
                            }
                          }
                          type="text"
                          value="B1, Channel 1"
                        />
                      </div>
                      <div
                        className="col-sm-3"
                        style={
                          Object {
                            "paddingLeft": "0px",
                          }
                        }
                      >
                        <input
                          className="form-control input-small plan-input panel-plan-text"
                          id="plan-duration-1"
                          maxLength={8}
                          minLength={1}
                          onChange={[Function]}
                          onFocus={[Function]}
                          placeholder="Duration"
                          style={
                            Object {
                              "maxWidth": "100%",
                            }
                          }
                          type="text"
                          value="30"
                        />
                      </div>
                      <label
                        className="panel-plan-text"
                        style={
                          Object {
                            "marginLeft": "-15px",
                            "paddingRight": "5px",
                          }
                        }
                      >
                        s
                      </label>
                      <label
                        className="col-sm-1 radio-btn"
                      >
                        <input
                          checked={true}
                          id="plan-row-1"
                          onChange={[Function]}
                          style={
                            Object {
                              "maxWidth": "100%",
                            }
                          }
                          type="radio"
                        />
                        <span
                          className="checkmark"
                        />
                      </label>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </ReactList>
        </form>
      </div>
      <div
        className="container"
      >
        <div
          className="row text-center btn-row-1"
        >
          <div
            className="col-sm-3"
          >
            <button
              className="btn btn-block btn-group"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              Insert
            </button>
          </div>
          <div
            className="col-sm-3"
          >
            <button
              className="btn btn-block btn-light btn-group"
              onClick={[Function]}
              type="button"
            >
              Save
            </button>
          </div>
          <div
            className="col-sm-3"
          >
            <button
              className="btn btn-block btn-light btn-group"
              onClick={[Function]}
              type="button"
            >
              Load
            </button>
          </div>
        </div>
        <div
          className="row btn-row-2"
        >
          <div
            className="col-sm-3"
          >
            <button
              className="btn btn-block btn-cancel btn-group"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              Delete
            </button>
          </div>
          <div
            className="col-sm-3"
          >
            <button
              className="btn btn-block btn-cancel btn-group"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              Clear
            </button>
          </div>
          <div
            className="col-sm-3"
          >
            <button
              className="btn btn-block btn-green btn-group"
              onClick={[Function]}
              type="button"
            >
              OK
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</PlanPanel>
`;

exports[`<PlanPanel /> Snapshot 1`] = `
<div>
  <div
    className="panel-plan"
  >
    <h2
      className="panel-plan-text"
    >
      Schedule
    </h2>
    <span
      className="cancel panel-plan-text"
      onClick={[Function]}
    />
    <h6
      className="panel-plan-text"
    >
      Please click on available channels to set up a schedule, then click on the radio button to select starting position.
    </h6>
    <div>
      <h6
        className="panel-plan-text"
      >
        Currently viewing File:
         
        <span
          style={
            Object {
              "color": "white",
            }
          }
        >
          test.pln
        </span>
      </h6>
    </div>
    <div
      className="panel-plan-inner"
    >
      <form>
        <ReactList
          axis="y"
          itemRenderer={[Function]}
          itemsRenderer={[Function]}
          length={32}
          minSize={1}
          pageSize={10}
          scrollParentGetter={[Function]}
          scrollParentViewportSizeGetter={[Function]}
          threshold={100}
          type="uniform"
          useStaticSize={false}
          useTranslate3d={false}
        />
      </form>
    </div>
    <div
      className="container"
    >
      <div
        className="row text-center btn-row-1"
      >
        <div
          className="col-sm-3"
        >
          <button
            className="btn btn-block btn-group"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Insert
          </button>
        </div>
        <div
          className="col-sm-3"
        >
          <button
            className="btn btn-block btn-light btn-group"
            onClick={[Function]}
            type="button"
          >
            Save
          </button>
        </div>
        <div
          className="col-sm-3"
        >
          <button
            className="btn btn-block btn-light btn-group"
            onClick={[Function]}
            type="button"
          >
            Load
          </button>
        </div>
      </div>
      <div
        className="row btn-row-2"
      >
        <div
          className="col-sm-3"
        >
          <button
            className="btn btn-block btn-cancel btn-group"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Delete
          </button>
        </div>
        <div
          className="col-sm-3"
        >
          <button
            className="btn btn-block btn-cancel btn-group"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Clear
          </button>
        </div>
        <div
          className="col-sm-3"
        >
          <button
            className="btn btn-block btn-green btn-group"
            onClick={[Function]}
            type="button"
          >
            OK
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<PlanPanel /> makePlanRow 1`] = `
<div
  className="gf-form-inline"
>
  <label
    className="col-sm-1 panel-plan-text"
    style={
      Object {
        "fontSize": "17px",
        "marginTop": "5px",
      }
    }
  >
    5. 
  </label>
  <div
    className="col-sm-6 col-bank-name"
  >
    <input
      className="form-control plan-input panel-plan-text"
      id="plan-port-5"
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="Select port"
      style={
        Object {
          "float": "left",
          "marginLeft": "2px",
          "maxWidth": "90%",
        }
      }
      type="text"
      value=""
    />
  </div>
  <div
    className="col-sm-3"
    style={
      Object {
        "paddingLeft": "0px",
      }
    }
  >
    <input
      className="form-control input-small plan-input panel-plan-text"
      id="plan-duration-5"
      maxLength={8}
      minLength={1}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="Duration"
      style={
        Object {
          "maxWidth": "100%",
        }
      }
      type="text"
      value=""
    />
  </div>
  <label
    className="panel-plan-text"
    style={
      Object {
        "marginLeft": "-15px",
        "paddingRight": "5px",
      }
    }
  >
    s
  </label>
  <label
    className="col-sm-1 radio-btn"
  >
    <input
      checked={false}
      id="plan-row-5"
      onChange={[Function]}
      style={
        Object {
          "maxWidth": "100%",
        }
      }
      type="radio"
    />
    <span
      className="checkmark"
    />
  </label>
</div>
`;

exports[`<PlanPanel /> renderItem 1`] = `<div />`;
